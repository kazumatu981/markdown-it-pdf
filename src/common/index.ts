export * from './configure';
export * from './logger';
