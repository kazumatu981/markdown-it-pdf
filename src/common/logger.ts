import chalk = require('chalk');

/**
 * Represents a logging method.
 */
type LogMethods = 'trace' | 'debug' | 'info' | 'warn' | 'error';

const methodIndexes = ['trace', 'debug', 'info', 'warn', 'error'];

/**
 * Represents a log level.
 */
export type LogLevel = LogMethods | 'silent';

/**
 * Represents a log level index.
 */
export const levelIndexes = [
    'trace',
    'debug',
    'info',
    'warn',
    'error',
    'silent',
];

/**
 * Represents a logger interface.
 */
export interface Logger {
    /**
     * Logs a trace message.
     * @param message - The message to log.
     */
    trace: (message: unknown, ...args: unknown[]) => void;
    /**
     * Logs a debug message.
     * @param message - The message to log.
     */
    debug: (message: unknown, ...args: unknown[]) => void;
    /**
     * Logs an info message.
     * @param message - The message to log.
     */
    info: (message: unknown, ...args: unknown[]) => void;
    /**
     * Logs a warning message.
     * @param message - The message to log.
     */
    warn: (message: unknown, ...args: unknown[]) => void;
    /**
     * Logs an error message.
     * @param message - The message to log.
     */
    error: (message: unknown, ...args: unknown[]) => void;
}

/**
 * Implements a logger that logs messages to the console.
 */
export class ConsoleLogger implements Logger {
    private level: LogLevel;

    /**
     * Implements a logger that logs messages to the console.
     * @param level {LogLevel} The log level.
     */
    public constructor(level: LogLevel = 'info') {
        this.level = level;
    }
    /**
     * Logs a trace message.
     * @param message - The message to log.
     * @param args - Additional arguments to pass to the log function.
     */
    public trace(message: unknown, ...args: unknown[]): void {
        // Logs a trace message at the trace log level.
        this.log('trace', message, ...args);
    }
    /**
     * Logs a debug message.
     * @param message - The message to log.
     * @param args - Additional arguments to pass to the log function.
     */
    public debug(message: unknown, ...args: unknown[]): void {
        // Logs a debug message at the debug log level.
        this.log('debug', message, ...args);
    }
    /**
     * Logs an info message.
     * @param message - The message to log.
     * @param args - Additional arguments to pass to the log function.
     */
    public info(message: unknown, ...args: unknown[]): void {
        // Logs an info message at the info log level.
        this.log('info', message, ...args);
    }
    /**
     * Logs a warning message.
     * @param message - The message to log.
     * @param args - Additional arguments to pass to the log function.
     */
    public warn(message: unknown, ...args: unknown[]) {
        // Logs a warning message at the warn log level.
        this.log('warn', message, ...args);
    }
    /**
     * Logs an error message.
     * @param message - The message to log.
     * @param args - Additional arguments to pass to the log function.
     */
    public error(message: unknown, ...args: unknown[]) {
        // Logs an error message at the error log level.
        this.log('error', message, ...args);
    }

    // #region Private Methods
    /**
     * Determines if the specified log level should be logged.
     * @param level - The log level to check.
     * @returns True if the log level should be logged, false otherwise.
     */
    private shouldLog(level: LogLevel): boolean {
        // Get the index of the log level in the list of log levels.
        const levelIndex = methodIndexes.indexOf(level);

        // Get the index of the current log level in the list of log levels.
        const currentLevelIndex = methodIndexes.indexOf(this.level);

        // If the log level index is greater than or equal to the current log level index,
        // then the log level should be logged.
        return levelIndex >= currentLevelIndex;
    }

    /**
     * Creates a log message with the specified log level and message.
     * @param {LogLevel} level - The log level.
     * @param {string} message - The message to log.
     * @returns {string} The log message.
     */
    private createMessage(level: LogLevel, message: unknown): string {
        // Format the log message with the log level, timestamp, and message.
        let levelString: string = level;
        switch (level) {
            case 'trace':
                levelString = chalk.yellowBright(level.toUpperCase());
                break;
            case 'debug':
                levelString = chalk.gray(level.toUpperCase());
                break;
            case 'info':
                levelString = chalk.blue(level.toUpperCase());
                break;
            case 'warn':
                levelString = chalk.redBright(level.toUpperCase());
                break;
            case 'error':
                levelString = chalk.bgRed(level.toUpperCase());
                break;
        }
        const messageString = `[${levelString} ${getCurrentTimeString()}] ${message}`;

        // Return the formatted log message.
        switch (level) {
            case 'trace':
            case 'debug':
                return chalk.gray(messageString);
            default:
                return messageString;
        }
    }

    /**
     * Logs a message at the specified log level.
     * If the log level is lower than the current log level, the message is not logged.
     * @param level The log level to log the message at.
     * @param message The message to log.
     * @param args Any additional arguments to pass to the log function.
     */
    private log(level: LogMethods, message: unknown, ...args: unknown[]) {
        // If the log level is set to 'silent', do not log anything.
        if (this.level === 'silent') return;

        // Get the correct log function based on the log level.
        const logCallBack = console[level];

        // If the log level is high enough to be logged, log the message.
        if (this.shouldLog(level)) {
            // Create the log message with the log level and timestamp.
            const logMessage = this.createMessage(level, message);

            // Log the message using the appropriate log function.
            logCallBack(logMessage, ...args);
        }
    }
    // #endregion
}

/**
 * Returns a string representing the current time in the format "HH:MM:SS".
 * @returns {string} The current time in the format "HH:MM:SS".
 */
function getCurrentTimeString(): string {
    const date = new Date();
    return `${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}`;
}
