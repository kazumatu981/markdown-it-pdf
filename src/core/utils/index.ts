export * from './http-helper';
export * from './path-resolver';
