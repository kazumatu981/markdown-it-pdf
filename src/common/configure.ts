import fs from 'fs';
import { type Logger } from './logger';
import { type PDFOptions } from 'puppeteer';
import { type ContentsMapOptions } from '../core/maps/contents-map';

export interface Range {
    min?: number;
    max?: number;
}
export interface ServerPortOptions {
    retry?: number;
    range?: Range;
}

export interface MarkdownRenderServerOptions
    extends ContentsMapOptions,
        ServerPortOptions {
    port?: number;
    externalUrls?: string[];
}

export type PrinterOptions = Omit<PDFOptions, 'path'>;

export interface MarkdownItPdfRenderServerOptions
    extends MarkdownRenderServerOptions {}
export interface MarkdownItPdfPrinterOptions
    extends MarkdownItPdfRenderServerOptions {
    printerOption?: PrinterOptions;
    outputDir?: string;
}

export function readOptions<T>(
    filePath?: string,
    logger?: Logger
): T | undefined {
    let options: T | undefined = undefined;
    if (filePath && fs.existsSync(filePath)) {
        try {
            if (filePath.endsWith('.json')) {
                options = readJsonOptions<T>(filePath);
            } else if (filePath.endsWith('.js') || filePath.endsWith('.cjs')) {
                options = readJSOptions<T>(filePath);
            } else {
                logger?.warn(
                    'Unsupported configuration file extension: %s, so using default options.',
                    filePath
                );
            }
        } catch (_) {
            // nothing to do
            logger?.warn(
                'Failed to read configuration file: %s, so using default options.',
                filePath
            );
        }
    } else {
        logger?.info(
            'Configuration file not found: %s, so using default options.',
            filePath
        );
    }
    return options;
}

function readJsonOptions<T>(filePath: string): T | undefined {
    const content = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(content) as T;
}

function readJSOptions<T>(filePath: string): T | undefined {
    const module = require(filePath);
    return module as T;
}
