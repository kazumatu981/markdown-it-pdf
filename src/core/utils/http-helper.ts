import http from 'http';
import { type Logger } from '../../common';

//#region constants
const privatePortRange = {
    min: 49152,
    max: 65535,
};

const defaultListeningOptions = {
    retry: 10,
    range: privatePortRange,
};
//#endregion

//#region types and interfaces

/**
 * The server that is listening on a given port
 */
interface ListeningHttpServer {
    /**
     * The port that the server is listening on.
     */
    port: number;
    /**
     * The http server that is listening on the specified port.
     */
    httpServer: http.Server;
}

/**
 * The range of ports to search for
 */
export interface Range {
    /**
     * The minimum port to search for
     */
    min?: number;
    /**
     * The maximum port to search for
     */
    max?: number;
}

/**
 * Options for listening on a port
 */
export interface ListeningOptions {
    /**
     * The number of times to retry if the port is in use
     */
    retry?: number;
    /**
     * The range of ports to search for
     */
    range?: Range;
}

//#endregion

//#region functions

//#region exported functions
/**
 * Tries to listen on a given port. If the port is not specified, it tries to
 * listen on a random port within a specified range. If the specified port is
 * in use, it retries the specified number of times.
 *
 * @param {number | undefined} port The port to listen on. If not specified, it will try to listen on
 * a random port within the specified range.
 * @param {ListeningOptions | undefined} options The options for listening on a port.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
export async function tryToListen(
    port?: number,
    options?: ListeningOptions,
    logger?: Logger
): Promise<ListeningHttpServer | undefined> {
    let listeningHttpServer = undefined;
    if (port != undefined) {
        logger?.debug('Try to Listen on port %d', port);
        listeningHttpServer = await tryToListenCore(port, logger);
    } else {
        logger?.debug('Try to Listen on port random mode');
        for (
            let i = 0;
            i < (options?.retry ?? defaultListeningOptions.retry);
            i++
        ) {
            port = getRandom(options?.range);
            listeningHttpServer = await tryToListenCore(port, logger);
            if (listeningHttpServer !== undefined) {
                break; // success
            }
        }
    }
    return listeningHttpServer;
}
//#endregion

//#region internal functions
/**
 * Generates a random number within a specified range.
 *
 * @param {Range | undefined} range - The range within which the random number
 * should be generated. If not specified, the function will use the default range
 * of 49152 to 65535.
 * @returns {number} A random number within the specified range.
 * @throws {Error} If the range is invalid (min > max).
 */
function getRandom(range?: Range): number {
    // Set the default range if not specified
    const candidate = {
        min: range?.min ?? defaultListeningOptions.range.min,
        max: range?.max ?? defaultListeningOptions.range.max,
    };

    // Check if the range is valid
    if (candidate.min > candidate.max) {
        throw new Error('Invalid range');
    }

    // Generate a random number within the range
    return Math.floor(
        Math.random() * (candidate.max - candidate.min) + candidate.min
    );
}

/**
 * Tries to listen on a given port. If the port is in use, it rejects the
 * promise.
 *
 * @param {number} port The port to listen on.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
function tryToListenCore(
    port: number,
    logger?: Logger
): Promise<ListeningHttpServer | undefined> {
    // Create a new http server
    const server = http.createServer();

    // Create a promise that resolves to the port that the server is listening on
    // or undefined if the specified port is in use and cannot be listened on
    return new Promise((resolve) => {
        // Start listening on the specified port
        server.listen(port, () => {
            // Log the reserved port
            logger?.debug('reserved port: %d', port);
            // Resolve the promise with the port and the server
            resolve({
                port,
                httpServer: server,
            });
        });
        // If there is an error while listening on the port, log the error and
        // resolve the promise with undefined
        server.on('error', (error) => {
            logger?.debug('listen error: %s', error.message);
            logger?.trace(error.stack);
            logger?.warn('port %d may be used.', port);
            resolve(undefined);
        });
    });
}
//#endregion

//#endregion
