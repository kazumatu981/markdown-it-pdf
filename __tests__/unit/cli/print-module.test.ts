import { jest, expect, describe, it } from '@jest/globals';
import { ConsoleLogger } from '../../../src/common/logger';
import { mockLogger } from '../../utils/mock-logger';
import printModule from '../../../src/cli/print-module';
import { mockingTestDir, unmockingTestDir } from '../../utils/test-dir';
jest.mock('../../../src/common/logger', () => ({
    ConsoleLogger: jest.fn(() => {
        return mockLogger;
    }),
}));

describe('yargs module unit tests - print-module', () => {
    afterEach(() => {
        jest.resetAllMocks();
    });
    afterAll(() => {
        jest.restoreAllMocks();
    });
    it('handler', async () => {
        mockingTestDir();
        await printModule.handler({
            dir: 'test',
            outputDir: 'pdf',
        });
        unmockingTestDir();
        expect(mockLogger.info).toMatchSnapshot();
    });
});
