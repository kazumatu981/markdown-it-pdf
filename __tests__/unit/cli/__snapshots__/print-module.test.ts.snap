// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yargs module unit tests - print-module builder 1`] = `
[
  "dir",
  "outputDir",
]
`;

exports[`yargs module unit tests - print-module handler - on error 1`] = `
[
  "Error Occurred on printing: %s",
]
`;

exports[`yargs module unit tests - print-module handler 1`] = `
[
  "MarkdownItPDF Printer is starting...",
  "ready to print.",
  "%d files printed",
]
`;

exports[`yargs module unit tests - print-module handler 2`] = `
[
  "createPdfPrinter() called with options: {"rootDir":"test","outputDir":"pdf","port":3001}",
  "MarkdownRenderServer.createRenderServer() called with options: {"rootDir":"test","outputDir":"pdf","port":3001}",
  "contentsUrls: %o",
  "Try to Listen on port %d",
  "reserved port: %d",
  "Listening on port %d",
  "Launched Puppeteer.",
  "Request: %s",
  "render() called.",
  "styles: %o",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Request: %s",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Request: %s",
  "Printed http://localhost:3001/test.md to pdf\\test.md.pdf.",
  "Request: %s",
  "render() called.",
  "styles: %o",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Request: %s",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Printed http://localhost:3001/sub/test.md to pdf\\sub\\test.md.pdf.",
  "Request: %s",
  "render() called.",
  "styles: %o",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Request: %s",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Printed http://localhost:3001/sub2/test.md to pdf\\sub2\\test.md.pdf.",
  "Request: %s",
  "render() called.",
  "styles: %o",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Request: %s",
  "Response for url: %s, status code: %d",
  "content type: %s",
  "Printed http://localhost:3001/sub2/sub/test.md to pdf\\sub2\\sub\\test.md.pdf.",
  "Closed Puppeteer.",
  "printed files's ids:",
  [
    "/test.md",
    "/sub/test.md",
    "/sub2/test.md",
    "/sub2/sub/test.md",
  ],
]
`;
